# Исследование надёжности заёмщиков
____
### Это предварительное исследование, результаты которого будут учтены при построении модели крединтого скоринга. Заказчик — кредитный отдел банка. Основные задачи предобработать данные и выявить влияние семейного положения и количества детей клиента на факт погашения кредита в срок. <a class="tocSkip">
____
##### Даны ответы на вопросы: <a class="tocSkip">
 - Есть ли зависимость между наличием детей и возвратом кредита в срок?
 - Есть ли зависимость между семейным положением и возвратом кредита в срок?
 - Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
 - Как разные цели кредита влияют на его возврат в срок?
____
##### Входные данные от банка — статистика о платёжеспособности клиентов. <a class="tocSkip">

<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1">Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1">Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2">Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-дубликатов" data-toc-modified-id="Удаление-дубликатов-2.1">Удаление дубликатов</a></span></li><li><span><a href="#Удаление-наименее-представленных-типов-занятости" data-toc-modified-id="Удаление-наименее-представленных-типов-занятости-2.2">Удаление наименее представленных типов занятости</a></span></li><li><span><a href="#Замена-ошибочных-значений-в-количестве-детей" data-toc-modified-id="Замена-ошибочных-значений-в-количестве-детей-2.3">Замена ошибочных значений в количестве детей</a></span></li><li><span><a href="#Приведение-категорий-образования-к-единому-виду" data-toc-modified-id="Приведение-категорий-образования-к-единому-виду-2.4">Приведение категорий образования к единому виду</a></span></li><li><span><a href="#Приведение-категорий-семейного-статуса-к-единому-виду" data-toc-modified-id="Приведение-категорий-семейного-статуса-к-единому-виду-2.5">Приведение категорий семейного статуса к единому виду</a></span></li><li><span><a href="#Удаление-пропущенного-значения-пола-клиента" data-toc-modified-id="Удаление-пропущенного-значения-пола-клиента-2.6">Удаление пропущенного значения пола клиента</a></span></li><li><span><a href="#Перепроверка-и-удаление-дубликатов" data-toc-modified-id="Перепроверка-и-удаление-дубликатов-2.7">Перепроверка и удаление дубликатов</a></span></li><li><span><a href="#Добавление-целочисленных-категорий-наличия-и-количества-детей" data-toc-modified-id="Добавление-целочисленных-категорий-наличия-и-количества-детей-2.8">Добавление целочисленных категорий наличия и количества детей</a></span></li><li><span><a href="#Добавление-целочисленных-категорий-дохода" data-toc-modified-id="Добавление-целочисленных-категорий-дохода-2.9">Добавление целочисленных категорий дохода</a></span></li><li><span><a href="#Добавление-целочисленных-категорий-типов-занятости" data-toc-modified-id="Добавление-целочисленных-категорий-типов-занятости-2.10">Добавление целочисленных категорий типов занятости</a></span></li><li><span><a href="#Добавление-целочисленных-категорий-целей-кредита" data-toc-modified-id="Добавление-целочисленных-категорий-целей-кредита-2.11">Добавление целочисленных категорий целей кредита</a></span></li><li><span><a href="#Добавление-целочисленных-категорий-пола" data-toc-modified-id="Добавление-целочисленных-категорий-пола-2.12">Добавление целочисленных категорий пола</a></span></li><li><span><a href="#Заполнение-нулевых-значений-возраста" data-toc-modified-id="Заполнение-нулевых-значений-возраста-2.13">Заполнение нулевых значений возраста</a></span></li><li><span><a href="#Заполнение-пропущенных-значений-трудового-стажа" data-toc-modified-id="Заполнение-пропущенных-значений-трудового-стажа-2.14">Заполнение пропущенных значений трудового стажа</a></span></li><li><span><a href="#Заполнение-пропущенных-значений-дохода" data-toc-modified-id="Заполнение-пропущенных-значений-дохода-2.15">Заполнение пропущенных значений дохода</a></span></li><li><span><a href="#Итоговая-проверка-и-корректировка-пропусков/дубликатов/типов-данных" data-toc-modified-id="Итоговая-проверка-и-корректировка-пропусков/дубликатов/типов-данных-2.16">Итоговая проверка и корректировка пропусков/дубликатов/типов данных</a></span></li><li><span><a href="#Вывод-" data-toc-modified-id="Вывод--2.17">Вывод <a id="finding_2"></a></a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы" data-toc-modified-id="Ответы-на-вопросы-3">Ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-3.1">Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2">Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3">Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4">Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4">Общий вывод</a></span></li></ul></div>

## Подготовка данных

In [1]:
import pandas as pd
from pymystem3 import Mystem

pd.set_option('max_columns', None)
pd.set_option('display.max_colwidth', 500)
pd.options.display.float_format = '{:.3f}'.format

In [2]:
df = pd.read_csv('data.csv')
df.sample(10)

children  days_employed  dob_years education  education_id  \
17129         1       -986.911         31   среднее             1   
7259          0     -11556.539         57    высшее             0   
13199         0      -4906.386         52    ВЫСШЕЕ             0   
14826         2      -1430.589         42   среднее             1   
11459         0      -8067.465         55   среднее             1   
4100          0      -4157.418         35   среднее             1   
13483         0       -810.927         61   среднее             1   
13481         0      -1858.367         37   среднее             1   
12313         0        -84.875         43   среднее             1   
2094          0     363129.470         51   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
17129        женат / замужем                 0      M    сотрудник     0   
7259         женат / замужем                 0      F    сотрудник     0   
13199       гражданский брак                 1      F    компаньон     0   
14826        женат / замужем                 0      F    сотрудник     0   
11459        женат / замужем                 0      M  госслужащий     0   
4100         женат / замужем                 0      M    компаньон     0   
13483        женат / замужем                 0      F  госслужащий     0   
13481  Не женат / не замужем                 4      F    сотрудник     1   
12313  Не женат / не замужем                 4      F    сотрудник     0   
2094               в разводе                 3      F    пенсионер     0   

       total_income                            purpose  
17129     97057.060                    свой автомобиль  
7259     201414.907                       недвижимость  
13199    139863.801                    сыграть свадьбу  
14826    121373.415                 высшее образование  
11459     88705.743  покупка коммерческой недвижимости  
4100     255420.220                       недвижимость  
13483    328607.232              заняться образованием  
13481    107900.433                         автомобиль  
12313     83460.027            покупка жилья для сдачи  
2094     161181.192               покупка недвижимости

In [3]:
display(df.info())
display(df.duplicated().sum())
display(df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


None

54

children  days_employed  dob_years  education_id  family_status_id  \
count 21525.000      19351.000  21525.000     21525.000         21525.000   
mean      0.539      63046.498     43.293         0.817             0.973   
std       1.382     140827.312     12.575         0.548             1.420   
min      -1.000     -18388.950      0.000         0.000             0.000   
25%       0.000      -2747.424     33.000         1.000             0.000   
50%       0.000      -1203.370     42.000         1.000             0.000   
75%       1.000       -291.096     53.000         1.000             1.000   
max      20.000     401755.400     75.000         4.000             4.000   

           debt  total_income  
count 21525.000     19351.000  
mean      0.081    167422.302  
std       0.273    102971.566  
min       0.000     20667.264  
25%       0.000    103053.153  
50%       0.000    145017.938  
75%       0.000    203435.068  
max       1.000   2265604.029

In [4]:
for col in df.columns:
    display(df[col].value_counts(dropna=False))

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

NaN          2174
-1580.623       1
-4122.461       1
-2828.238       1
-2636.091       1
             ... 
-7120.518       1
-2146.884       1
-881.455        1
-794.666        1
-3382.114       1
Name: days_employed, Length: 19352, dtype: int64

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

0    19784
1     1741
Name: debt, dtype: int64

NaN           2174
69374.480        1
106196.236       1
248730.171       1
45654.214        1
              ... 
101516.605       1
239154.168       1
165009.733       1
94270.050        1
189255.287       1
Name: total_income, Length: 19352, dtype: int64

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

### Вывод
- **Описание данных:**
    - *children* - количество детей в семье;
    - *days_employed* - общий трудовой стаж в днях;
    - *dob_years* - возраст клиента в годах;
    - *education* - уровень образования клиента;
    - *education_id* - идентификатор уровня образования;
    - *family_status* - семейное положение;
    - *family_status_id* - идентификатор семейного положения;
    - *gender* - пол клиента;
    - *income_type* - тип занятости;
    - *debt* - имел ли задолженность по возврату кредитов;
    - *total_income* - ежемесячный доход;
    - *purpose* - цель получения кредита.


- В столбцах *days_employed* и *total_income* выявлены явные пропуски в количестве 2174 в каждом, что больше 10%.
- Выявлены полные дубликаты в количестве 54.


- После более детального изучения выявлены следующие проблемы:
    - *children* - 76 клиентов указали "20" детей, наиболее вероятно это ошибка заполнения форм, когда форма не очищалась от "0" по умолчанию. Подтверждает лишний ноль и отсутствие значений от 5 до 20. По причине подобной ошибки, полагаю, 47 клиентов указали '-1' детей. 
    - *days_employed* - кроме большого количества пропусков, выявленно большое количество отрицательных значений. После изучения вопроса понял, что это оценка с учётом коэффициентов.
    - *dob_years* - выявлены неявные пропуски - нулевые значения.
    - *education* - различия по регистру внутри категорий.
    - *family_status* - в одной из категорий разный регистр символов.
    - *gender* - нет данных о поле одного клиента.
    - *income_type* - четыре типа представлены всего в шести строках датасета;
    - *purpose* - различное написание одинаковых по смыслу категорий.

## Предобработка данных

**План предобработки:**
1. удалить 54 дублирующихся строки;
2. *income_type* - удалить четыре наименее представленных в датасете типа занятости;
3. *children*:
    - 20 заменить на 2;
    - -1 заменить на 1;
4. *education* - привести категории к единому виду, проверить сравнив с *education_id*;
5. *family_status* - привести категории к единому виду, проверить сравнив с *family_status_id*;
6. *gender* - удалить строку с пропуском;
7. добавить столбец с целочисленными категориями наличия и количества детей - нет детей - 0 , один ребёнок - 1, два и более - 2;
8. добавить столбец с целочисленными категориями дохода - низкий(первый квартиль) - 0, средний(два интерквартильных размаха после первого квартиля) - 1, высокий(больше двух интерквартильных размахов после первого квартиля) - 2;
9. добавить столбец с целочисленными категориями типов занятости:
     - сотрудник - 0;
     - компаньон - 1;
     - пенсионер - 2;
     - госслужащий - 3;
10. добавить столбец с целочисленными категориями целей кредита - свадьба - 0, недвижимость - 1, авто - 2, образование - 3;
11. добавить столбец с целочисленными категориями пола - F - 0, M - 1;
12. *dob_years* -  найти связь с другими признаками и заполнить нулевые значения;
13. *days_employed* - найти связь с другими признаками и заполнить пропуски;
14. *total_income*  - найти связь с другими признаками и заполнить пропуски;
15. проверить пропуски в столбцах с категориями и заполнить.

### Удаление дубликатов

In [5]:
print('Дубликатов до удаления:', df.duplicated().sum())
df.drop_duplicates(inplace=True)
print('Дубликатов после удаления:', df.duplicated().sum())

Дубликатов до удаления: 54
Дубликатов после удаления: 0


### Удаление наименее представленных типов занятости

In [6]:
display(df['income_type'].value_counts(dropna=False))
print(
    'Удаляемые 4 из 8 типов занятости представляют {:.3%} датасета'
    .format(df['income_type'].value_counts(dropna=False).tail(4).sum() / len(df['income_type']))
)
low_types = df['income_type'].value_counts(dropna=False).tail(4).index
df.query('income_type not in @low_types', inplace=True)
display(df['income_type'].value_counts(dropna=False))

сотрудник          11091
компаньон           5080
пенсионер           3837
госслужащий         1457
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

Удаляемые 4 из 8 типов занятости представляют 0.028% датасета


сотрудник      11091
компаньон       5080
пенсионер       3837
госслужащий     1457
Name: income_type, dtype: int64

### Замена ошибочных значений в количестве детей

In [7]:
display(df['children'].value_counts())
df['children'].replace({20: 2, -1: 1}, inplace=True)
display(df['children'].value_counts())

 0     14103
 1      4808
 2      2051
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

0    14103
1     4855
2     2127
3      330
4       41
5        9
Name: children, dtype: int64

### Приведение категорий образования к единому виду

In [8]:
display(df.groupby('education_id')['education'].unique())
df['education'] = df['education'].str.lower()
display(df.groupby('education_id')['education'].unique())

education_id
0                                           [высшее, ВЫСШЕЕ, Высшее]
1                                        [среднее, Среднее, СРЕДНЕЕ]
2    [неоконченное высшее, НЕОКОНЧЕННОЕ ВЫСШЕЕ, Неоконченное высшее]
3                                  [начальное, НАЧАЛЬНОЕ, Начальное]
4                   [Ученая степень, УЧЕНАЯ СТЕПЕНЬ, ученая степень]
Name: education, dtype: object

education_id
0                 [высшее]
1                [среднее]
2    [неоконченное высшее]
3              [начальное]
4         [ученая степень]
Name: education, dtype: object

### Приведение категорий семейного статуса к единому виду

In [9]:
display(df.groupby('family_status_id')['family_status'].unique())
df['family_status'] = df['family_status'].str.lower()
display(df.groupby('family_status_id')['family_status'].unique())

family_status_id
0          [женат / замужем]
1         [гражданский брак]
2           [вдовец / вдова]
3                [в разводе]
4    [Не женат / не замужем]
Name: family_status, dtype: object

family_status_id
0          [женат / замужем]
1         [гражданский брак]
2           [вдовец / вдова]
3                [в разводе]
4    [не женат / не замужем]
Name: family_status, dtype: object

### Удаление пропущенного значения пола клиента

In [10]:
display(df['gender'].value_counts(dropna=False))
df.drop(df[df['gender'] == 'XNA'].index[0], inplace=True)
display(df['gender'].value_counts(dropna=False))

F      14186
M       7278
XNA        1
Name: gender, dtype: int64

F    14186
M     7278
Name: gender, dtype: int64

### Перепроверка и удаление дубликатов

In [11]:
print('Дубликатов возникло после предобработки:', df.duplicated().sum())
df.drop_duplicates(inplace=True)
print('Дубликатов после удаления:',  df.duplicated().sum())

Дубликатов возникло после предобработки: 17
Дубликатов после удаления: 0


### Добавление целочисленных категорий наличия и количества детей

In [12]:
def children_category(row):
    if row['children'] == 0:
        return 0
    if row['children'] == 1:
        return 1
    return 2
   
df['children_cat'] = df.apply(children_category, axis=1)
df.groupby('children_cat')['children'].unique()

children_cat
0             [0]
1             [1]
2    [3, 2, 4, 5]
Name: children, dtype: object

### Добавление целочисленных категорий дохода

In [13]:
display(df['total_income'].describe())
low = df['total_income'].describe()[4]
medium = low + (df['total_income'].describe()[6] - low)*2

def total_income_category(row):
    if pd.isna(row['total_income']):
        return row['total_income']
    if row['total_income'] <= low:
        return 0
    if row['total_income'] <= medium:
        return 1
    return 2
   
df['total_income_cat'] = df.apply(total_income_category, axis=1)
df['total_income_cat'].astype('int64', copy=False, errors='ignore')
display(df.groupby('total_income_cat')['total_income'].describe().T)

count     19345.000
mean     167416.448
std      102951.921
min       20667.264
25%      103082.408
50%      145017.938
75%      203425.992
max     2265604.029
Name: total_income, dtype: float64

total_income_cat      0.000      1.000       2.000
count              4837.000  13071.000    1437.000
mean              78312.223 172740.131  418920.466
std               17384.228  50535.713  167975.876
min               20667.264 103092.791  303783.321
25%               66285.144 131469.389  331312.025
50%               81298.908 162344.369  370135.024
75%               92686.195 205553.549  441478.520
max              103082.408 303518.271 2265604.029

### Добавление целочисленных категорий типов занятости

In [14]:
display(df['income_type'].value_counts(dropna=False))

def income_type_category(row):
    if row['income_type'] == 'госслужащий':
        return 3
    if row['income_type'] == 'пенсионер':
        return 2
    if row['income_type'] == 'компаньон':
        return 1
    return 0
   
df['income_type_cat'] = df.apply(income_type_category, axis=1)
display(df.groupby('income_type_cat')['income_type'].unique())

сотрудник      11084
компаньон       5077
пенсионер       3829
госслужащий     1457
Name: income_type, dtype: int64

income_type_cat
0      [сотрудник]
1      [компаньон]
2      [пенсионер]
3    [госслужащий]
Name: income_type, dtype: object

### Добавление целочисленных категорий целей кредита

In [15]:
frame_for_lemmas = df['purpose'].value_counts(dropna=False).to_frame().reset_index()

m = Mystem()

def lemmatisizer(row):
    return ' '.join(m.lemmatize(row['index'])) 
 
frame_for_lemmas['lemmatisized'] = frame_for_lemmas.apply(lemmatisizer, axis = 1)
frame_for_lemmas

index  purpose  \
0                                  свадьба      791   
1                    на проведение свадьбы      767   
2                          сыграть свадьбу      765   
3                 операции с недвижимостью      675   
4        покупка коммерческой недвижимости      661   
5                        операции с жильем      652   
6    операции с коммерческой недвижимостью      650   
7                  покупка жилья для сдачи      650   
8                                    жилье      646   
9                            покупка жилья      646   
10                 покупка жилья для семьи      638   
11  строительство собственной недвижимости      634   
12                            недвижимость      633   
13         операции со своей недвижимостью      627   
14        строительство жилой недвижимости      624   
15                    покупка своего жилья      620   
16                    покупка недвижимости      620   
17              строительство недвижимости      619   
18                            ремонт жилью      606   
19              покупка жилой недвижимости      605   
20            на покупку своего автомобиля      505   
21            заняться высшим образованием      496   
22                              автомобиль      493   
23        сделка с подержанным автомобилем      486   
24                              автомобили      478   
25                         свой автомобиль      478   
26      на покупку подержанного автомобиля      478   
27                   на покупку автомобиля      471   
28                 приобретение автомобиля      461   
29              дополнительное образование      460   
30                    сделка с автомобилем      455   
31                      высшее образование      452   
32                             образование      447   
33   получение дополнительного образования      446   
34                   получение образования      442   
35                  профильное образование      436   
36           получение высшего образования      426   
37                   заняться образованием      408   

                                     lemmatisized  
0                                      свадьба \n  
1                    на   проведение   свадьба \n  
2                            сыграть   свадьба \n  
3                  операция   с   недвижимость \n  
4        покупка   коммерческий   недвижимость \n  
5                         операция   с   жилье \n  
6   операция   с   коммерческий   недвижимость \n  
7                покупка   жилье   для   сдача \n  
8                                        жилье \n  
9                              покупка   жилье \n  
10               покупка   жилье   для   семья \n  
11  строительство   собственный   недвижимость \n  
12                                недвижимость \n  
13         операция   со   свой   недвижимость \n  
14        строительство   жилой   недвижимость \n  
15                      покупка   свой   жилье \n  
16                      покупка   недвижимость \n  
17                строительство   недвижимость \n  
18                              ремонт   жилье \n  
19              покупка   жилой   недвижимость \n  
20            на   покупка   свой   автомобиль \n  
21          заниматься   высокий   образование \n  
22                                  автомобиль \n  
23       сделка   с   подержанный   автомобиль \n  
24                                  автомобиль \n  
25                           свой   автомобиль \n  
26       на   покупка   подержать   автомобиль \n  
27                   на   покупка   автомобиль \n  
28                   приобретение   автомобиль \n  
29                дополнительный   образование \n  
30                     сделка   с   автомобиль \n  
31                       высокий   образование \n  
32                                 образование \n  
33    получение   дополнительный   образование \n  
34                     получение   образование \n  
35                 

In [16]:
def purpose_category(row):
    if 'автомобиль' in row['lemmatisized']:
        return 2
    if 'свадьба' in row['lemmatisized']:
        return 0
    if 'образование' in row['lemmatisized']:
        return 3
    return 1

frame_for_lemmas['purpose_cat'] = frame_for_lemmas.apply(purpose_category, axis = 1)

frame_for_lemmas

index  purpose  \
0                                  свадьба      791   
1                    на проведение свадьбы      767   
2                          сыграть свадьбу      765   
3                 операции с недвижимостью      675   
4        покупка коммерческой недвижимости      661   
5                        операции с жильем      652   
6    операции с коммерческой недвижимостью      650   
7                  покупка жилья для сдачи      650   
8                                    жилье      646   
9                            покупка жилья      646   
10                 покупка жилья для семьи      638   
11  строительство собственной недвижимости      634   
12                            недвижимость      633   
13         операции со своей недвижимостью      627   
14        строительство жилой недвижимости      624   
15                    покупка своего жилья      620   
16                    покупка недвижимости      620   
17              строительство недвижимости      619   
18                            ремонт жилью      606   
19              покупка жилой недвижимости      605   
20            на покупку своего автомобиля      505   
21            заняться высшим образованием      496   
22                              автомобиль      493   
23        сделка с подержанным автомобилем      486   
24                              автомобили      478   
25                         свой автомобиль      478   
26      на покупку подержанного автомобиля      478   
27                   на покупку автомобиля      471   
28                 приобретение автомобиля      461   
29              дополнительное образование      460   
30                    сделка с автомобилем      455   
31                      высшее образование      452   
32                             образование      447   
33   получение дополнительного образования      446   
34                   получение образования      442   
35                  профильное образование      436   
36           получение высшего образования      426   
37                   заняться образованием      408   

                                     lemmatisized  purpose_cat  
0                                      свадьба \n            0  
1                    на   проведение   свадьба \n            0  
2                            сыграть   свадьба \n            0  
3                  операция   с   недвижимость \n            1  
4        покупка   коммерческий   недвижимость \n            1  
5                         операция   с   жилье \n            1  
6   операция   с   коммерческий   недвижимость \n            1  
7                покупка   жилье   для   сдача \n            1  
8                                        жилье \n            1  
9                              покупка   жилье \n            1  
10               покупка   жилье   для   семья \n            1  
11  строительство   собственный   недвижимость \n            1  
12                                недвижимость \n            1  
13         операция   со   свой   недвижимость \n            1  
14        строительство   жилой   недвижимость \n            1  
15                      покупка   свой   жилье \n            1  
16                      покупка   недвижимость \n            1  
17                строительство   недвижимость \n            1  
18                              ремонт   жилье \n            1  
19              покупка   жилой   недвижимость \n            1  
20            на   покупка   свой   автомобиль \n            2  
21          заниматься   высокий   образование \n            3  
22                                  автомобиль \n            2  
23       сделка   с   подержанный   автомобиль \n            2  
24                                  автомобиль \n            2  
25                           свой   автомобиль \n            2  
26       на   покупка   подержать   автомобиль \n            2  
27                   на   покупка   автомобиль \n            2  
28    

In [17]:
purpose_category_dict = frame_for_lemmas.groupby('index')['purpose_cat'].mean().to_dict()

def purpose_category_for_df(row):
    return purpose_category_dict[row['purpose']]

df['purpose_cat'] = df.apply(purpose_category_for_df, axis=1)
df.groupby('purpose_cat')['purpose'].unique().to_frame()

purpose
purpose_cat                                                                                                                                                                                                                                                                                                                                                                                                                                            
0                                                                                                                                                                                                                                                                                                                                                                                                     [сыграть свадьбу, на проведение свадьбы, свадьба]
1            [покупка жилья, операции с жильем, покупка жилья для семьи, покупка недвижимости, покупка коммерческой недвижимости, покупка жилой недвижимости, строительство собственной недвижимости, недвижимость, строительство недвижимости, операции с коммерческой недвижимостью, строительство жилой недвижимости, жилье, операции со своей недвижимостью, покупка своего жилья, операции с недвижимостью, покупка жилья для сдачи, ремонт жилью]
2                                                                                                                                                                                                                                   [приобретение автомобиля, на покупку подержанного автомобиля, на покупку своего автомобиля, автомобили, сделка с подержанным автомобилем, автомобиль, свой автомобиль, сделка с автомобилем, на покупку автомобиля]
3                                                                                                                                                                                                               [дополнительное образование, образование, заняться образованием, получение образования, получение дополнительного образования, получение высшего образования, профильное образование, высшее образование, заняться высшим образованием]

### Добавление целочисленных категорий пола

In [18]:
display(df['gender'].value_counts(dropna=False))

def gender_category(row):
    if 'F' in row['gender']:
        return 0
    return 1

df['gender_cat'] = df.apply(gender_category, axis = 1)

display(df['gender_cat'].value_counts(dropna=False))

F    14171
M     7276
Name: gender, dtype: int64

0    14171
1     7276
Name: gender_cat, dtype: int64

### Заполнение нулевых значений возраста

In [19]:
def find_corr_for_col(col):
    display(df.corr()[col].abs().sort_values(ascending=False))

In [20]:
find_corr_for_col('dob_years')

dob_years          1.000
days_employed      0.583
income_type_cat    0.351
children_cat       0.320
children           0.309
gender_cat         0.130
family_status_id   0.070
debt               0.069
total_income_cat   0.069
education_id       0.067
total_income       0.053
purpose_cat        0.009
Name: dob_years, dtype: float64

In [21]:
display(df['dob_years'].describe())

years_dict = df.groupby(['income_type_cat', 'children_cat', 'gender_cat'])['dob_years'].median().to_dict()

def fill_zero_age(row):
    if row['dob_years'] == 0:
        return years_dict.get((row['income_type_cat'], row['children_cat'], row['gender_cat']))
    return row['dob_years']

df['dob_years'] = df.apply(fill_zero_age, axis = 1)

display(df['dob_years'].describe())

count   21447.000
mean       43.274
std        12.570
min         0.000
25%        33.000
50%        42.000
75%        53.000
max        75.000
Name: dob_years, dtype: float64

count   21447.000
mean       43.480
std        12.227
min        19.000
25%        33.500
50%        43.000
75%        53.000
max        75.000
Name: dob_years, dtype: float64

13. *days_employed* - найти связь с другими признаками и заполнить пропуски;
14. *total_income*  - найти связь с другими признаками и заполнить пропуски;
15. проверить пропуски в столбцах с категориями и заполнить.

### Заполнение пропущенных значений трудового стажа

In [22]:
find_corr_for_col('days_employed')

days_employed      1.000
dob_years          0.603
income_type_cat    0.580
children_cat       0.245
children           0.236
total_income_cat   0.156
gender_cat         0.152
total_income       0.137
education_id       0.081
debt               0.047
purpose_cat        0.007
family_status_id   0.006
Name: days_employed, dtype: float64

In [23]:
display(df['days_employed'].describe())

days_dict = df.groupby(['dob_years', 'income_type_cat'])['days_employed'].median().to_dict()

def fill_na_days(row):
    if pd.isna(row['days_employed']):
        return days_dict.get((row['dob_years'], row['income_type_cat']))
    return row['days_employed']

df['days_employed'] = df.apply(fill_na_days, axis = 1)

display(df['days_employed'].describe())

count    19345.000
mean     63028.519
std     140811.983
min     -18388.950
25%      -2747.549
50%      -1203.507
75%       -291.133
max     401755.400
Name: days_employed, dtype: float64

count    21444.000
mean     63247.217
std     140907.801
min     -18388.950
25%      -2581.174
50%      -1283.656
75%       -317.424
max     401755.400
Name: days_employed, dtype: float64

### Заполнение пропущенных значений дохода

In [24]:
find_corr_for_col('total_income')

total_income       1.000
total_income_cat   0.735
gender_cat         0.180
education_id       0.179
days_employed      0.137
dob_years          0.056
children           0.032
children_cat       0.031
income_type_cat    0.018
debt               0.012
purpose_cat        0.010
family_status_id   0.009
Name: total_income, dtype: float64

In [25]:
display(df['total_income'].describe())

income_dict = df.groupby(['gender_cat', 'education_id'])['total_income'].median().to_dict()

def fill_na_income(row):
    if pd.isna(row['total_income']):
        return income_dict.get((row['gender_cat'], row['education_id']))
    return row['total_income']

df['total_income'] = df.apply(fill_na_income, axis = 1)

display(df['total_income'].describe())

count     19345.000
mean     167416.448
std      102951.921
min       20667.264
25%      103082.408
50%      145017.938
75%      203425.992
max     2265604.029
Name: total_income, dtype: float64

count     21447.000
mean     165417.430
std       98259.109
min       20667.264
25%      107633.565
50%      144869.450
75%      198097.128
max     2265604.029
Name: total_income, dtype: float64

### Итоговая проверка и корректировка пропусков/дубликатов/типов данных

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21447 entries, 0 to 21524
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21447 non-null  int64  
 1   days_employed     21444 non-null  float64
 2   dob_years         21447 non-null  int64  
 3   education         21447 non-null  object 
 4   education_id      21447 non-null  int64  
 5   family_status     21447 non-null  object 
 6   family_status_id  21447 non-null  int64  
 7   gender            21447 non-null  object 
 8   income_type       21447 non-null  object 
 9   debt              21447 non-null  int64  
 10  total_income      21447 non-null  float64
 11  purpose           21447 non-null  object 
 12  children_cat      21447 non-null  int64  
 13  total_income_cat  19345 non-null  float64
 14  income_type_cat   21447 non-null  int64  
 15  purpose_cat       21447 non-null  int64  
 16  gender_cat        21447 non-null  int64 

In [27]:
def fill_na_total_income_category(row):
    if pd.isna(row['total_income_cat']):
        if pd.isna(row['total_income']):
            return row['total_income']
        if row['total_income'] <= low:
            return 0
        if row['total_income'] <= medium:
            return 1
        return 2
    return row['total_income_cat']
   
df['total_income_cat'] = df.apply(fill_na_total_income_category, axis=1)
display(df.groupby('total_income_cat')['total_income'].describe().T)

total_income_cat      0.000      1.000       2.000
count              4837.000  15173.000    1437.000
mean              78312.223 169177.005  418920.466
std               17384.228  48587.393  167975.876
min               20667.264 103092.791  303783.321
25%               66285.144 127226.816  331312.025
50%               81298.908 158983.316  370135.024
75%               92686.195 200001.773  441478.520
max              103082.408 303518.271 2265604.029

In [28]:
df.dropna(inplace=True)
df = df.astype('int64', errors='ignore')

In [29]:
display(df.info())
display(df.duplicated().sum())
display(df.describe())
for col in df.columns:
    display(df[col].value_counts(dropna=False))

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21444 entries, 0 to 21524
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21444 non-null  int64 
 1   days_employed     21444 non-null  int64 
 2   dob_years         21444 non-null  int64 
 3   education         21444 non-null  object
 4   education_id      21444 non-null  int64 
 5   family_status     21444 non-null  object
 6   family_status_id  21444 non-null  int64 
 7   gender            21444 non-null  object
 8   income_type       21444 non-null  object
 9   debt              21444 non-null  int64 
 10  total_income      21444 non-null  int64 
 11  purpose           21444 non-null  object
 12  children_cat      21444 non-null  int64 
 13  total_income_cat  21444 non-null  int64 
 14  income_type_cat   21444 non-null  int64 
 15  purpose_cat       21444 non-null  int64 
 16  gender_cat        21444 non-null  int64 
dtypes: int64(12)

None

0

children  days_employed  dob_years  education_id  family_status_id  \
count 21444.000      21444.000  21444.000     21444.000         21444.000   
mean      0.481      63247.541     43.481         0.817             0.974   
std       0.756     140907.421     12.226         0.549             1.422   
min       0.000     -18388.000     19.000         0.000             0.000   
25%       0.000      -2580.500     34.000         1.000             0.000   
50%       0.000      -1283.000     43.000         1.000             0.000   
75%       1.000       -317.000     53.000         1.000             1.000   
max       5.000     401755.000     75.000         4.000             4.000   

           debt  total_income  children_cat  total_income_cat  \
count 21444.000     21444.000     21444.000         21444.000   
mean      0.081    165420.758         0.460             0.841   
std       0.273     98265.233         0.694             0.517   
min       0.000     20667.000         0.000             0.000   
25%       0.000    107629.000         0.000             1.000   
50%       0.000    144875.000         0.000             1.000   
75%       0.000    198109.500         1.000             1.000   
max       1.000   2265604.000         2.000             2.000   

       income_type_cat  purpose_cat  gender_cat  
count        21444.000    21444.000   21444.000  
mean             0.797        1.467       0.339  
std              0.962        0.916       0.473  
min              0.000        0.000       0.000  
25%              0.000        1.000       0.000  
50%              0.000        1.000       0.000  
75%              1.000        2.000       1.000  
max              3.000        3.000       1.000

0    14084
1     4854
2     2126
3      330
4       41
5        9
Name: children, dtype: int64

-1740      71
-1361      47
-1487      42
-1957      42
-1849      42
           ..
-3273       1
 400581     1
-5308       1
-3257       1
 376824     1
Name: days_employed, Length: 9105, dtype: int64

40    622
35    621
34    607
41    605
38    597
42    596
33    581
39    574
44    565
31    557
36    556
37    555
29    544
30    537
48    536
43    521
50    513
32    509
49    508
28    503
45    495
27    492
52    484
56    483
47    477
54    476
46    474
53    459
57    456
58    453
51    446
55    443
59    443
26    408
60    393
25    357
61    355
62    348
63    269
24    262
64    260
23    252
65    193
22    182
66    182
67    167
21    111
68     99
69     84
70     65
71     56
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

среднее                15168
высшее                  5245
неоконченное высшее      743
начальное                282
ученая степень             6
Name: education, dtype: int64

1    15168
0     5245
2      743
3      282
4        6
Name: education_id, dtype: int64

женат / замужем          12335
гражданский брак          4147
не женат / не замужем     2808
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

0    12335
1     4147
4     2808
3     1195
2      959
Name: family_status_id, dtype: int64

F    14168
M     7276
Name: gender, dtype: int64

сотрудник      11084
компаньон       5076
пенсионер       3827
госслужащий     1457
Name: income_type, dtype: int64

0    19705
1     1739
Name: debt, dtype: int64

125635    997
158983    481
162830    364
204544    169
142725     44
         ... 
179138      1
61724       1
99269       1
151371      1
126820      1
Name: total_income, Length: 18603, dtype: int64

свадьба                                   790
на проведение свадьбы                     767
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
операции с коммерческой недвижимостью     650
покупка жилья для сдачи                   650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    634
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      620
строительство недвижимости                619
покупка своего жилья                      619
ремонт жилью                              606
покупка жилой недвижимости                605
на покупку своего автомобиля              505
заняться высшим образованием      

0    14084
1     4854
2     2506
Name: children_cat, dtype: int64

1    15170
0     4837
2     1437
Name: total_income_cat, dtype: int64

0    11084
1     5076
2     3827
3     1457
Name: income_type_cat, dtype: int64

1    10805
2     4304
3     4013
0     2322
Name: purpose_cat, dtype: int64

0    14168
1     7276
Name: gender_cat, dtype: int64

### Вывод <a id='finding_2'></a>

 - После предобработки потеряна 81 строка. Это меньше половины процента датасета.
 - Устранены пропуски и дубликаты.
 - Данные подготовлены для последующего анализа и подготовки модели крединтого скоринга.

## Ответы на вопросы

### Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [30]:
children_pivot = df.pivot_table(index='children_cat', values='debt', aggfunc='mean')
children_pivot.rename(columns={'debt':'Вероятность просрочки'}, inplace=True)
children_pivot = children_pivot.T
children_pivot.rename(columns={0:'Нет детей', 1:'Один ребёнок', 2:'Два и более'}, inplace=True)
children_pivot

children_cat           Нет детей  Один ребёнок  Два и более
Вероятность просрочки      0.075         0.091        0.093

Зависимость между наличием детей и возвратом кредита наблюдается. Бездетные реже имеют просрочки по кредитам - 7,5% против 9,1% у заёмщиков с одним ребёнком и 9,3% с двумя и более.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [31]:
family_pivot = (
    df
    .pivot_table(index='family_status', values='debt', aggfunc='mean')
    .sort_values('debt', ascending=True)
)
family_pivot.rename(columns={'debt':'Вероятность просрочки'}, inplace=True)
family_pivot = family_pivot.T
family_pivot

family_status          вдовец / вдова  в разводе  женат / замужем  \
Вероятность просрочки           0.066      0.071            0.075   

family_status          гражданский брак  не женат / не замужем  
Вероятность просрочки             0.094                  0.098

Зависимость между семейным положением и возвратом кредита наблюдается. Вдовцы и вдовы, а так же разведённые реже имеют просрочки по кредитам - 6,6% и 7,1% соответственно. За ними следуют женатые/замужние, которые имеют просрочки по кредитам в 7,5% случаев. Чаще других имеют просрочки по кредитам состоящие в гражданском браке и неженатые/незамужние - 9,4% и 9,8% соответственно.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [32]:
income_pivot = (
    df
    .pivot_table(index='total_income_cat', values='debt', aggfunc='mean')
    .sort_values('debt', ascending=True)
)
income_pivot.rename(columns={'debt':'Вероятность просрочки'}, inplace=True)
income_pivot = income_pivot.T
income_pivot.rename(columns={0:'Низкий', 1:'Средний', 2:'Высокий'}, inplace=True)
income_pivot

total_income_cat       Высокий  Низкий  Средний
Вероятность просрочки    0.070   0.079    0.083

Зависимость между уровнем дохода и возвратом кредита наблюдается. Клиенты с высоким доходом реже имеют просрочки по кредитам - 7%. За ними следуют клиенты с низким доходом которые имеют просрочки по кредитам в 7,9% случаев. Чаще других имеют просрочки по кредитам клиенты со средним доходом - 8,3%.

### Как разные цели кредита влияют на его возврат в срок?

In [33]:
purpose_pivot = (
    df
    .pivot_table(index='purpose_cat', values='debt', aggfunc='mean')
    .sort_values('debt', ascending=True)
)
purpose_pivot.rename(columns={'debt':'Вероятность просрочки'}, inplace=True)
purpose_pivot = purpose_pivot.T
purpose_pivot.rename(columns={0:'Свадьба', 1:'Недвижимость', 2:'Авто', 3:'Образование'}, inplace=True)
purpose_pivot

purpose_cat            Недвижимость  Свадьба  Образование  Авто
Вероятность просрочки         0.072    0.080        0.092 0.093

Кредиты взятые на различные операции с недвижимостью реже просрачиваются - 7,2%. За ними следуют кредиты взятые на свадебные мероприятия - 8%. Значительно чаще просрочки возникают на образовательные кредиты - 9,2%. Автокредиты самые частые по просрочкам - 9,3%.

## Общий вывод

*Если просто: бездетные богатые вдовы/вдовцы берущие кредит на недвижимость реже всех имеют просрочки по кредитам.*

 - Зависимость между наличием детей и возвратом кредита наблюдается. Бездетные реже имеют просрочки по кредитам - 7,5% против 9,1% у заёмщиков с одним ребёнком и 9,3% с двумя и более.
 - Зависимость между семейным положением и возвратом кредита наблюдается. Вдовцы и вдовы, а так же разведённые реже имеют просрочки по кредитам - 6,6% и 7,1% соответственно. За ними следуют женатые/замужние, которые имеют просрочки по кредитам в 7,5% случаев. Чаще других имеют просрочки по кредитам состоящие в гражданском браке и неженатые/незамужние - 9,4% и 9,8% соответственно.
 - Зависимость между уровнем дохода и возвратом кредита наблюдается. Клиенты с высоким доходом реже имеют просрочки по кредитам - 7%. За ними следуют клиенты с низким доходом которые имеют просрочки по кредитам в 7,9% случаев. Чаще других имеют просрочки по кредитам клиенты со средним доходом - 8,3%.
 - Кредиты взятые на различные операции с недвижимостью реже просрачиваются - 7,2%. За ними следуют кредиты взятые на свадебные мероприятия - 8%. Значительно чаще просрочки возникают на образовательные кредиты - 9,2%. Автокредиты самые частые по просрочкам - 9,3%.